java/vector/src/main/java/org/apache/arrow/vector/complex/impl/PromotableWriter.java

```diff
@@ -118,21 +126,22 @@ public void setPosition(int index) {
     }
   }
 
-  protected FieldWriter getWriter(MinorType type) {
+  @Override
+  protected FieldWriter getWriter(ArrowType type) {
     if (state == State.UNION) {
-      ((UnionWriter)writer).getWriter(type);
+      ((UnionWriter) writer).getWriter(Types.getMinorTypeForArrowType(type));
```

Comment:
UnionWriter doesn't work for Decimal it looks like (https://github.com/apache/arrow/blob/master/java/vector/src/main/codegen/templates/UnionWriter.java#L124)
It's fine not to support FixedWidthList in union either. (at least for now)